### DISCUSSION TITLE
Issue with geographic to topocetric projection

### DISCUSSION THREAD

**Discussion by yuvalg1987** [pyproj4/pyproj#863] [Q&A]
Hi, 
I'm trying to move from a geographic projection into a topocentric one. I found the following proj4 line that does the trick but can't seem to get it running using pyproj. What am I missing? how should I create this new CRS? 

BTW I'm new to proj4/pyproj so I might just misunderstanding the basic concept. I'm trying to use geopandas to_crs() method using the proj4 line for the ocnversion. 

```
+proj=pipeline \
    +step +proj=cart +ellps=WGS84 \
    +step +proj=topocentric +ellps=WGS84 +lon_0=5 +lat_0=55 +h_0=200

```
Thanks
Yuval

**Comment by snowman2** [comment]
See: https://github.com/OSGeo/PROJ/pull/2444

Do you have PROJ 8+?

**Reply by yuvalg1987** [reply]
Thanks for the quick reply! I've already visited this link so I know this change is relatively new.
I'm using pyproj version 3.0.1 and proj version 8.0.1

One thing i'm a bit confused about is the difference between a CRS and a Transformer/Coordinate Operation. Geopandas requires a CRS input in order to transform from one coordinate system to another (any thing that can be used within pyproj.CRS.from_user_input()). Unfortunately the following code crashes:

```
enu_proj =  "+proj=pipeline " \
                "+step +proj=cart +ellps=WGS84" \
                "+step +proj=topocentric +ellps=WGS84 +lon_0=5 +lat_0=55 +h_0=200"

    new_crs = CRS(enu_proj)
```
On the other hand, creating an operation doesn't (although I have no idea how to continue from this stage)
```
    ENU_OP =  CoordinateOperation.from_string(
        "+proj=pipeline "
        "+step +proj=cart +ellps=WGS84 "
        "+step +proj=topocentric +lat_0=35.01908 +lon_0=32.83495 +h_0=0 +ellps=WGS84"
    ) 
```
I think the problem is in my understanding of the working principals of proj rather than a code issue.

**Reply by snowman2** [reply]
I could be wrong, but I don't think `geopandas` will be able to help in this scenario (until something like this is added: https://github.com/geopandas/geopandas/issues/1175) as I believe you need to use `Transformer.from_pipeline` (#713): https://pyproj4.github.io/pyproj/stable/api/transformer.html#pyproj.transformer.Transformer.from_pipeline.

**Reply by snowman2** [reply]
Actually, this might work:
```python
src_crs = CRS("+proj=cart +ellps=WGS84")
dst_crs = CRS("+proj=topocentric +ellps=WGS84 +lon_0=5 +lat_0=55 +h_0=200")
gdf.set_crs(src_crs)
gdf2 = gdf.to_crs(dst_crs)
```

**Reply by yuvalg1987** [reply]
Thanks for the suggestion. I'll try it asap. What alternative to geopandas do I have for reading a .shp file and performing this type of operation?

**Reply by yuvalg1987** [reply]
Ok, using the Transformer seems to work correctly and you're probably right regarding [geopandas/geopandas#1175](https://github.com/geopandas/geopandas/issues/1175) so I guess geopandas is not the way to go for this. Thanks for the help!

**Reply by snowman2** [reply]
@yuvalg1987 does this work with geopandas?
```python
src_crs = CRS("+proj=cart +ellps=WGS84")
dst_crs = CRS("+proj=topocentric +ellps=WGS84 +lon_0=5 +lat_0=55 +h_0=200")
gdf.set_crs(src_crs)
gdf2 = gdf.to_crs(dst_crs)
```

**Reply by yuvalg1987** [reply]
No, Didn't work. I think it has to do with the fact that this is a custom combination that cannot be solved by the default configuration of PROJ4.

**Reply by snowman2** [reply]
That's unfortunate. Good luck :+1:

**Comment by stefangachter** [comment]
Dear all, I am also stuck with the topocentric conversion. I am using pyproj version 3.1.0, assuming that the above mentioned pull request #2444 is included. If I follow the example, https://proj.org/operations/conversions/topocentric.html,
```
import pyproj
pipe_trans = pyproj.Transformer.from_pipeline("+proj=pipeline +step +proj=cart +ellps=WGS84 +step +proj=topocentric +ellps=WGS84 +lon_0=5 +lat_0=55 +h_0=200")
```
then I get an error
`
ProjError("Invalid projection b'+proj=pipeline +step +proj=cart +ellps=WGS84 +step +proj=topocentric +ellps=WGS84 +lon_0=5 +lat_0=55 +h_0=200'.: (Internal Proj Error: proj_create: Error -5 (unknown projection id): Pipeline: Bad step definition: proj=topocentric (unknown projection id))")
`
indicating that topocentric conversion is unknown as a projection. Do I have to define it not as a transformation? If yes, can you provide and example?

**Reply by stefangachter** [reply]
The topocentric conversion is part of version 8.0.1 https://proj.org/news.html#id28 (May 5, 2021) pyproj is version 3.1.0 (May 24, 2021). It might contain the topocentric conversion, but I am not sure.

**Reply by stefangachter** [reply]
I tested with the latest docker
`PS C:\WINDOWS\system32> docker run osgeo/proj cct -d 3 +proj=pipeline +step +proj=cart +ellps=WGS84 +step +proj=topocentric +ellps=WGS84 +lon_0=5 +lat_0=55 +h_0=200`
which seems to work, that is no error message, which suggest that the issue might be with pyproj version or how I setup the transformation command.

**Reply by snowman2** [reply]
Latest works:
```python
>>> import pyproj
>>> pipe_trans = pyproj.Transformer.from_pipeline("+proj=pipeline +step +proj=cart +ellps=WGS84 +step +proj=topocentric +ellps=WGS84 +lon_0=5 +lat_0=55 +h_0=200")
>>> pyproj.__proj_version__
'8.0.1'
>>> pyproj.__version__
'3.1.0'
>>> pipe_trans
<Other Coordinate Operation Transformer: pipeline>
Description: PROJ-based coordinate operation
Area of Use:
- undefined

```

**Reply by stefangachter** [reply]
Thanks a lot for the confirmation. I understand my problem now. There is a mismatch of the version:
```
>>> import pyproj
>>> pipe_trans = pyproj.Transformer.from_pipeline("+proj=pipeline +step +proj=cart +ellps=WGS84 +step +proj=topocentric +ellps=WGS84 +lon_0=5 +lat_0=55 +h_0=200")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\site-packages\pyproj\transformer.py", line 579, in from_pipeline
    return Transformer(TransformerFromPipeline(proj_pipeline))
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\site-packages\pyproj\transformer.py", line 311, in __init__
    self._local.transformer = transformer_maker()
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\site-packages\pyproj\transformer.py", line 125, in __call__
    return _Transformer.from_pipeline(cstrencode(self.proj_pipeline))
  File "pyproj\_transformer.pyx", line 696, in pyproj._transformer._Transformer.from_pipeline
pyproj.exceptions.ProjError: Invalid projection b'+proj=pipeline +step +proj=cart +ellps=WGS84 +step +proj=topocentric +ellps=WGS84 +lon_0=5 +lat_0=55 +h_0=200'.: (Internal Proj Error: proj_create: Error -5 (unknown projection id): Pipeline: Bad step definition: proj=topocentric (unknown projection id))
>>> pyproj.__proj_version__
'7.2.1'
>>> pyproj.__version__
'3.1.0'
>>>
```

**Reply by stefangachter** [reply]
In my case, pyproj version 3.1.0 refers to proj version 7.2.1, whereas it should be 8.0.1.

**Reply by stefangachter** [reply]
Is it because I am working on Windows? I installed it from https://pypi.org/project/pyproj/ as well as the wheels from https://www.lfd.uci.edu/~gohlke/pythonlibs/#pyproj, but I have always version 7.2.1
From https://pypi.org/project/pyproj/
```
PS C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64> pip install --no-cache-dir pyproj
Collecting pyproj
  Downloading pyproj-3.1.0-cp37-cp37m-win_amd64.whl (14.5 MB)
     |████████████████████████████████| 14.5 MB 6.4 MB/s
Requirement already satisfied: certifi in c:\program files (x86)\microsoft visual studio\shared\python37_64\lib\site-packages (from pyproj) (2020.6.20)
Installing collected packages: pyproj
Successfully installed pyproj-3.1.0
PS C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64> python
Python 3.7.8 (tags/v3.7.8:4b47a5b6ba, Jun 28 2020, 08:53:46) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyproj
>>> pyproj.__proj_version__
'7.2.1'
```
From https://www.lfd.uci.edu/~gohlke/pythonlibs/#pyproj (version 3.1.0)
```
PS C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64> pip install pyproj-3.1.0-cp37-cp37m-win_amd64.whl
Processing c:\program files (x86)\microsoft visual studio\shared\python37_64\pyproj-3.1.0-cp37-cp37m-win_amd64.whl
Requirement already satisfied: certifi in c:\program files (x86)\microsoft visual studio\shared\python37_64\lib\site-packages (from pyproj==3.1.0) (2020.6.20)
Installing collected packages: pyproj
Successfully installed pyproj-3.1.0
PS C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64> python
Python 3.7.8 (tags/v3.7.8:4b47a5b6ba, Jun 28 2020, 08:53:46) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> pyproj.__proj_version__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'pyproj' is not defined
>>> import pyproj
>>> pyproj.__proj_version__
'7.2.1'
```
From https://www.lfd.uci.edu/~gohlke/pythonlibs/#pyproj (version 3.0.1)
```
PS C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64> pip install .\pyproj-3.0.1-cp37-cp37m-win_amd64.whl
Processing c:\program files (x86)\microsoft visual studio\shared\python37_64\pyproj-3.0.1-cp37-cp37m-win_amd64.whl
Requirement already satisfied: certifi in c:\program files (x86)\microsoft visual studio\shared\python37_64\lib\site-packages (from pyproj==3.0.1) (2020.6.20)
Installing collected packages: pyproj
Successfully installed pyproj-3.0.1
PS C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64> python
Python 3.7.8 (tags/v3.7.8:4b47a5b6ba, Jun 28 2020, 08:53:46) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyproj
>>> pyproj.__proj_version__
'7.2.1'
```

**Reply by snowman2** [reply]
Interesting, that seems to be the case. Are you able to try installing with conda?

**Reply by stefangachter** [reply]
Do you mean that it is an issue of pyproj under Windows? If yes, how much time does it take to fix it? No, I haven’t tested with conda. Guidelines how to that?

**Reply by snowman2** [reply]
> Do you mean that it is an issue of pyproj under Windows?

It means that the Windows wheels have an older version of PROJ packaged inside. The easiest solution would be to wait for the next release coming soon.

For conda, see: https://pyproj4.github.io/pyproj/stable/installation.html

**Reply by stefangachter** [reply]
Thanks. I followed the instruction. I installed miniconda and pyproj with
`conda install -c conda-forge pyproj`
But then I run Anaconda's python, then I run into the following problem:
```
>>> import pyproj
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "D:\Documents\Programs\Miniconda3\lib\site-packages\pyproj\__init__.py", line 49, in <module>
    import pyproj.network
  File "D:\Documents\Programs\Miniconda3\lib\site-packages\pyproj\network.py", line 10, in <module>
    from pyproj._network import (  # noqa: F401
ImportError: DLL load failed while importing _network: 指定されたモジュールが見つかりません。
```
I followed the instructions at https://github.com/pyproj4/pyproj/issues/770 but installing the Microsoft Visual C++ Redistributables did not solve my issue.

**Reply by stefangachter** [reply]
Ah, it is about settings. If I using anaconda's python from within Visual Studio, then it cannot find the DLLs, but the using it from within Anaconda's power shell, then it is fine:

```
(base) PS C:\Users\gstef> python
Python 3.9.5 (default, May 18 2021, 14:42:02) [MSC v.1916 64 bit (AMD64)] :: Anaconda, Inc. on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyproj
>>> pyproj.__proj_version__
'8.1.0'
>>> pipe_trans = pyproj.Transformer.from_pipeline("+proj=pipeline +step +proj=cart +ellps=WGS84 +step +proj=topocentric +ellps=WGS84 +lon_0=5 +lat_0=55 +h_0=200")
>>> pipe_trans
<Other Coordinate Operation Transformer: pipeline>
Description: PROJ-based coordinate operation
Area of Use:
- undefined
>>>
```
Thus, with package from conda, the version is correct.

**Reply by stefangachter** [reply]
Ok, I could make it run with miniconda and Visual Studio 2019. Finally....
```
>>> import pyproj
>>> pyproj.__proj_version__
'8.1.0'
>>> pipe_trans = pyproj.Transformer.from_pipeline("+proj=pipeline +step +proj=cart +ellps=WGS84 +step +proj=topocentric +ellps=WGS84 +lon_0=5 +lat_0=55 +h_0=200")
>>> pipe_trans
<Other Coordinate Operation Transformer: pipeline>
Description: PROJ-based coordinate operation
Area of Use:
- undefined
>>> 
```
